### ISSUE TITLE
building Ganeti 2.11 with ghc 7.8

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1035]
Originally reported of Google Code with ID 979.

```
What software version are you running? Please provide the output of "gnt-
cluster --version", "gnt-cluster version", and "hspace --version".

<b>What distribution are you using?</b>
ArchLinux
ganeti-2.11 from source

<b>What steps will reproduce the problem?</b>
1. ./autogen.sh
2. ./configure [...]
3. make

<b>What is the expected output? What do you see instead?</b>
strerr:

src/htools.hs:37:1:
    Failed to load interface for ‘Ganeti.HTools.Program.Main’
    Use -v to see a list of the files searched for.
make: *** [src/htools.o] Error 1


<b>Please provide any additional information below.</b>
"tail" of stdout & stderr:

sed -e "s/%ver%/$VCSVER/" < src/Ganeti/Version.hs.in > src/Ganeti/Version.hs
/usr/bin/ghc -M -dep-makefile Makefile.ghc -dep-suffix hpc -O -Wall -Werror -isrc -fwarn-monomorphism-restriction -fwarn-tabs -fwarn-incomplete-uni-patterns    -itest/hs \
        -DPARALLEL3    src/htools.hs src/ganeti-kvmd.hs src/hluxid.hs src/hs2py.hs src/rpc-test.hs src/hconfd.hs src/ganeti-mond.hs src/mon-collector.hs 
[GHC]: src/htools.o <- 

src/htools.hs:37:1:
    Failed to load interface for ‘Ganeti.HTools.Program.Main’
    Use -v to see a list of the files searched for.
Makefile:3759: recipe for target 'src/htools.o' failed
make: *** [src/htools.o] Error 1


The File src/Ganeti/HTools/Program/Main.hs exists so I don't quite get how this error can happen
```
Originally added on 2014-10-18 00:21:17 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
same for Ganeti 2.12
```
Originally added on 2014-10-18 01:49:40 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-10-21 08:07:44 +0000 UTC.

Added Labels: Priority-High Type-Defect
Changed State: Accepted
Added to Milestone: Release2.11

**Comment by Ganeti-Issues-Migrator** [comment]
```
Switching priority to low, as we currently only support ghc 7.6.X
```
Originally added on 2014-10-21 13:44:07 +0000 UTC.

Added Labels: Priority-Low
Removed Labels: Priority-High

**Comment by Ganeti-Issues-Migrator** [comment]
```
It looks as if the dependencies in Makefile.ghc are generated wrongly and make is trying to compile htools.o straight away, without compiling the dependencies first. Could you please:

- add the type/version of 'make' you're using?

- check if patch [09ff6a25] helps (added to 2.11 just yesterday)
  http://git.ganeti.org/?p=ganeti.git;a=commit;h=09ff6a256470d525218f707181f43a7e628fda58

- attach the generated Makefile.ghc

Thanks.
```
Originally added on 2014-10-21 13:47:51 +0000 UTC.

Changed State: NeedFeedback

**Comment by Ganeti-Issues-Migrator** [comment]
```
make --version
GNU Make 4.0


yes, pulled just an hour ago, but no change
```
Originally added on 2014-10-21 14:06:42 +0000 UTC.

Attachments: 
[Makefile.ghc](https://storage.googleapis.com/google-code-attachments/ganeti/issue-979/comment-5/Makefile.ghc)

**Comment by Ganeti-Issues-Migrator** [comment]
```
I just compared the Makefile.ghc with one from debian stable and there is a lot missing


I then copied the Makefile.ghc from Debian over mine and tryed a make
this is the result:

[...]
src/Ganeti/BasicTypes.hs:212:24: Warning:
    In the use of type constructor or class ‘Error’
    (imported from Control.Monad.Error.Class, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

src/Ganeti/BasicTypes.hs:244:35: Warning:
    In the use of type constructor or class ‘Error’
    (imported from Control.Monad.Error.Class, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

src/Ganeti/BasicTypes.hs:246:42: Warning:
    In the use of ‘strMsg’
    (imported from Control.Monad.Error.Class, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

src/Ganeti/BasicTypes.hs:253:31: Warning:
    In the use of type constructor or class ‘Error’
    (imported from Control.Monad.Error.Class, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

src/Ganeti/BasicTypes.hs:254:26: Warning:
    In the use of ‘strMsg’
    (imported from Control.Monad.Error.Class, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

src/Ganeti/BasicTypes.hs:257:27: Warning:
    In the use of type constructor or class ‘Error’
    (imported from Control.Monad.Error.Class, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

src/Ganeti/BasicTypes.hs:264:26: Warning:
    In the use of type constructor or class ‘Error’
    (imported from Control.Monad.Error.Class, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

<no location info>: 
Failing due to -Werror.
Makefile:3770: recipe for target 'src/Ganeti/BasicTypes.o' failed
make: *** [src/Ganeti/BasicTypes.o] Error 1
```
Originally added on 2014-10-21 14:12:37 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
those Warnings (In the use of type constructor or class ‘Error’) I get as well if building on CentOS 7 (with ghc 7.6 & cabal builded dep.)


ghc --version
The Glorious Glasgow Haskell Compilation System, version 7.6.3
ganeti@node1 ~/src/ganeti-2.11 (git)-[stable-2.11] % make --version
GNU Make 3.82
Built for x86_64-redhat-linux-gnu
ganeti@node1 ~/src/ganeti-2.11 (git)-[stable-2.11] % uname -a
Linux node1.xxxxxxxxxxxx.de 3.10.0-123.el7.x86_64 #1 SMP Mon Jun 30 12:09:22 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

```
Originally added on 2014-10-21 22:20:54 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
with ganeti-2.11 & ArchLinux (ghc 7.8 & cabal) after patching


Loading package base ... linking ... done.

src/Ganeti/JSON.hs:107:18: Warning:
    In the use of type constructor or class ‘Error’
    (imported from Ganeti.BasicTypes, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

src/Ganeti/JSON.hs:108:43: Warning:
    In the use of ‘strMsg’
    (imported from Ganeti.BasicTypes, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

src/Ganeti/JSON.hs:206:15: Warning:
    In the use of type constructor or class ‘Error’
    (imported from Ganeti.BasicTypes, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"

src/Ganeti/JSON.hs:209:31: Warning:
    In the use of ‘strMsg’
    (imported from Ganeti.BasicTypes, but defined in Control.Monad.Trans.Error):
    Deprecated: "Use Control.Monad.Trans.Except instead"
[GHC]: src/Ganeti/THH.o <- src/Ganeti/THH.hs src/Ganeti/THH/PyType.hi src/Ganeti/THH/PyType.o src/Ganeti/PyValue.hi src/Ganeti/PyValue.o src/Ganeti/JSON.hi src/Ganeti/JSON.o

src/Ganeti/THH.hs:327:1: Warning: Defined but not used: ‘showJSONE’

src/Ganeti/THH.hs:354:1: Warning: Defined but not used: ‘appFn’

src/Ganeti/THH.hs:626:1: Warning:
    Defined but not used: type constructor or class ‘OpParam’

src/Ganeti/THH.hs:684:10: Warning:
    Defined but not used: ‘pyClassVar’

src/Ganeti/THH.hs:934:7: Warning: Defined but not used: ‘iname’

src/Ganeti/THH.hs:982:7: Warning: Defined but not used: ‘opid’
[GHC]: src/Ganeti/Logging.o <- src/Ganeti/Logging.hs src/Ganeti/ConstantUtils.hi src/Ganeti/ConstantUtils.o src/Ganeti/THH.hi src/Ganeti/THH.o
Loading package ghc-prim ... linking ... done.
Loading package integer-gmp ... linking ... done.
Loading package base ... linking ... done.
src/Ganeti/Logging.hs:1:1:
    cannot find normal object file ‘src/Ganeti/THH.dyn_o’
    while linking an interpreted expression
Makefile:3772: recipe for target 'src/Ganeti/Logging.o' failed
make: *** [src/Ganeti/Logging.o] Error 1
make  12.94s user 0.45s system 99% cpu 13.469 total

```
Originally added on 2014-10-22 19:56:57 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-10-27 12:37:29 +0000 UTC.

Changed State: New

**Comment by Ganeti-Issues-Migrator** [comment]
```
The `Deprecated: "Use Control.Monad.Trans.Except instead"` errors are because Ganeti (master, and older) is not yet -Werror compatible with transformers-0.4.

You can find the exact dependency versions I'm using for my GHC 7.8 build here: https://gist.github.com/nh2/d09eeac8400b32fb3288 (obtained with `cabal freeze`). We plan to release a cabal file in the future to make it clear with what dependency versions Ganeti works on what GHCs.

You can get around these warnings with the same fix as in 978 (on the stable-2.10 branch, -Werror is now only enforced with the `--developer-mode` and `--enable-haskell-tests` configure options), so if you build with stable-2.10 now, it should work.

Please tell me if this solves your issue!
```
Originally added on 2014-10-27 14:17:56 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-10-30 16:46:55 +0000 UTC.

Added Labels: Priority-Medium
Removed Labels: Priority-Low
Changed State: Accepted
Added to Milestone: Release2.13
Removed from Milestone: Release2.11

**Comment by Ganeti-Issues-Migrator** [comment]
```
The

    [GHC]: src/htools.o <- 
    
    src/htools.hs:37:1:
        Failed to load interface for ‘Ganeti.HTools.Program.Main’
        Use -v to see a list of the files searched for.
    Makefile:3759: recipe for target 'src/htools.o' failed
    make: *** [src/htools.o] Error 1

error is a result of https://ghc.haskell.org/trac/ghc/ticket/7381#comment:4.

This means we have to add `-dep-suffix ''` to the generation of `Makefile.ghc`.

I'll fix it.
```
Originally added on 2014-10-30 17:15:00 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
The "Failed to load interface" and follow-up errors are fixed in master with the commits.


commit 1ad14f38083d7d7702154f791070a92e241320e0
Author: Niklas Hambuechen <niklash@google.com>
Date:   Mon Nov 3 19:35:39 2014 +0100

    Makefile.am: Don't use -dynamic-too for .hpc_o files
    
    It turns out that GHC 7.8's -dynamic-too is forbidden not only for
    profiling builds object files, but also for those using HPC coverage.
    
    This commit accordingly renames HFLAGS_NOPROF to HFLAGS_DYNAMIC
    (since it now is not conditional on profiling any more),
    and makes sure that it is not used for profiling or HPC.
    
    This way we achieve that, for profiling+coverage builds
      - .dyn_o files are available for use in TH in the following 3 cases
      - .o files are for the normal binaries
      - .hpc_o files are for coverage-enabled test binaries
      - .prof_o files are for profiling binaries
    
    We make .hpc_o files depend on the .o files because the creation of
    the .o files will also create the .so files needed for TH.
    This was already in place for .prof_o files.
    
    This requires that HFLAGS now also contains `-itest/hs` since
    the rule for .o files is now also responsible for building .o/.so files
    in test/hs.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 9664aff98761f20fd8a037b01a44e78433d7c8c4
Author: Niklas Hambuechen <niklash@google.com>
Date:   Mon Nov 3 16:18:42 2014 +0100

    Makefile.am: Don't use dots in -osuf
    
    This fixes
      ./configure --enable-haskell-tests && make hs-tests
    builds on GHC 7.8.
    
    It changes `-osuf .ext` to `-osuf ext`; the old form was allowed in
    GHC <= 7.6, but GHC 7.8 seems to have problems with it:
      https://ghc.haskell.org/trac/ghc/ticket/9760
    
    The form without dot works in all GHC versions, and we don't have
    dots in other -osuf arguments in the same file either, so this
    was probably an accident.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 083776b9dbd7e704357841e6a8a4fce6802199fc
Author: Niklas Hambuechen <niklash@google.com>
Date:   Mon Nov 3 12:14:09 2014 +0100

    Fix compiler invocation for GHC >= 7.8
    
    GHC 7.8 switched to dynamic linking being used for ghci, which requires
    that .so file are being built for modules that are loaded for TH.
    
    For this reason, GHC >= 7.8 has a -dynamic-too flag, which we now use.
    
    However, -dynamic-too must not be enabled for profiling builds,
    which is why this commit introduces a HFLAGS_NOPROF variable
    (currently only containing -dynamic-too) that must be passed to
    all GHC invocations that are not creating profiling output.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

commit b78a2c3013c16395c48e055de82c4f93d9b41c38
Author: Niklas Hambuechen <niklash@google.com>
Date:   Fri Oct 31 16:40:27 2014 +0100

    Makefile.am: Fix wrong -dep-suffix for GHC 7.8
    
    This works around https://ghc.haskell.org/trac/ghc/ticket/9749:
    
    GHC 7.8 (undocumentedly) changed the way in which ghc -M generated
    object file dependencies, lacking the underscore that older versions
    added automatically to the file names.
    It also requires a -dep-suffix for the plain object file (.o).
    
    This commit detects GHC 7.6 and older (7.7 is development only, has no
    release and is treated equal to 7.8), and adjusts the -M invocation
    appropriately for newer GHC versions.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Reviewed-by: Hrvoje Ribicic <riba@google.com>

```
Originally added on 2014-11-05 11:01:32 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:37:03 +0000 UTC.

Changed State: Released
